File: packages/runtime-python/src/keyma/runtime/fields.py
"""Full field/ref set of a class across the inheritance chain (real inheritance).

``ClassMetadata["fields"]`` holds OWN fields only; inherited fields live on the ``base`` chain.
:func:`all_fields` assembles the complete set base-first (root → … → leaf), a child field
overriding an inherited one of the same name while keeping the ancestor's position — so JSON key
order, binary field order, and snapshots stay byte-stable. It mirrors the compiler's
``inheritedFields`` and the JS/C++ runtimes' ``allFields``/``all_fields``, preserving the
cross-runtime wire contract.

``refs`` is likewise own-only; :func:`all_refs` resolves embedded/reference targets by walking the
chain (leaf entries shadow ancestors'). Results are cached on the metadata dict's identity; class
metadata dicts are module-level singletons built once at import.
"""

from typing import Any, Dict, List, Optional

from .types import FieldMetadata, Metadata

_fields_cache: Dict[int, List[FieldMetadata]] = {}
_refs_cache: Dict[int, Dict[str, Any]] = {}


def all_fields(meta: Metadata) -> List[FieldMetadata]:
    base = meta.get("base")
    if base is None:
        return meta["fields"]
    key = id(meta)
    memo = _fields_cache.get(key)
    if memo is not None:
        return memo

    # Walk the base chain leaf-first (cycle-guarded by canonical ``name``).
    chain: List[Metadata] = []
    seen = set()
    cur: Optional[Metadata] = meta
    while cur is not None and cur.get("name") not in seen:
        seen.add(cur.get("name"))
        chain.append(cur)
        cur = cur.get("base")

    # Emit root-first; a dict keyed by field name gives each field the ancestor-position of its
    # first declaration while a child override supplies the winning definition.
    by_name: Dict[str, FieldMetadata] = {}
    for s in reversed(chain):
        for f in s["fields"]:
            by_name[f["name"]] = f
    result = list(by_name.values())
    _fields_cache[key] = result
    return result


def all_refs(meta: Metadata) -> Dict[str, Any]:
    base = meta.get("base")
    if base is None:
        return meta.get("refs") or {}
    key = id(meta)
    memo = _refs_cache.get(key)
    if memo is not None:
        return memo

    # Walk leaf → root; a leaf entry must win, so only set a key the first time it is seen.
    merged: Dict[str, Any] = {}
    seen = set()
    cur: Optional[Metadata] = meta
    while cur is not None and cur.get("name") not in seen:
        seen.add(cur.get("name"))
        for k, v in (cur.get("refs") or {}).items():
            if k not in merged:
                merged[k] = v
        cur = cur.get("base")
    _refs_cache[key] = merged
    return merged

File: packages/runtime/src/fields.ts
// Codec-facing metadata shape + field-chain helpers.
//
// IMPORTANT — the codec operates on the EMITTED data-model shape, not the legacy
// `SchemaMetadata`/`SchemaClass` declarations in `types.ts` (those survive only as the
// schema-domain generator's slicing source). A generated bundle's `<Class>.metadata` is a
// `ClassMetadata` whose reference/embedded field types carry `target` (not `schema`), whose
// brand exposes the metadata under `.metadata` (not `.schema`), and whose instances hydrate via
// the static `fromValue` factory (not `new Class(value)`). The structural types below mirror
// that emitted shape exactly, so the authored-here codec is baked verbatim and runs unchanged
// against generated classes.

/** A value type — the codec subset of the core IRType. `reference`/`embedded` carry the
 *  target class's canonical `name` under `target`; `instance` (param/return positions only,
 *  never a stored field) carries it under `name`. */
export type FieldType =
    | { kind: "string" }
    | { kind: "number"; bits?: 32 | 64 }
    | { kind: "integer"; bits?: 8 | 16 | 32 | 64; unsigned?: boolean }
    | { kind: "bigint" }
    | { kind: "boolean" }
    | { kind: "decimal" }
    | { kind: "bytes" }
    | { kind: "json" }
    | { kind: "date" }
    | { kind: "dateTime" }
    | { kind: "time" }
    | { kind: "id" }
    | { kind: "enum"; values: string[] }
    | { kind: "array"; of: FieldType; elementNullable?: boolean }
    | { kind: "reference"; target: string; idType?: FieldType }
    | { kind: "embedded"; target: string }
    | { kind: "instance"; name: string };

export type FieldMeta = {
    name: string;
    type: FieldType;
    /** Stable wire tag for binary serialization. Absent ⇒ the 1-based declaration index. */
    tag?: number;
};

/** The structural subset of a generated `<Class>.metadata` the codec reads. */
export type ClassMeta = {
    name: string;
    /** OWN fields only (real inheritance) — inherited fields live on `base`. */
    fields: FieldMeta[];
    /** Parent class's metadata when this class extends another; absent for a root. */
    base?: ClassMeta;
    /** Reference/embedded/instance target `name` → generated class. */
    refs?: ReadonlyMap<string, ClassRef>;
};

/** The structural subset of a generated model class the codec needs: its `metadata` brand and
 *  the static `fromValue` hydration factory. */
export interface ClassRef {
    readonly metadata: ClassMeta;
    fromValue(value: unknown): unknown;
}

// The full field set of a class, own + inherited (real inheritance). `ClassMeta.fields` holds
// OWN fields only; inherited fields live on the `base` chain. Assembled base-first (root → … →
// leaf), a child field overriding an inherited one of the same name while keeping the ancestor's
// position — identical to the order the compiler's `inheritedFields` (and the C++ `all_fields`)
// produce, so JSON key order and binary field order stay byte-stable across runtimes.
//
// Memoized: metadata objects are frozen singletons created once at module load.
const cache = new WeakMap<ClassMeta, readonly FieldMeta[]>();

export function allFields(meta: ClassMeta): readonly FieldMeta[] {
    if (meta.base === undefined) return meta.fields;
    const memo = cache.get(meta);
    if (memo !== undefined) return memo;

    // Walk the base chain leaf-first (cycle-guarded by canonical `name`).
    const chain: ClassMeta[] = [];
    const seen = new Set<string>();
    let cur: ClassMeta | undefined = meta;
    while (cur !== undefined && !seen.has(cur.name)) {
        seen.add(cur.name);
        chain.push(cur);
        cur = cur.base;
    }

    // Emit root-first; a Map keyed by field name gives each field the ancestor-position of its
    // first declaration while a child override supplies the winning definition.
    const byName = new Map<string, FieldMeta>();
    for (let i = chain.length - 1; i >= 0; i--) {
        for (const f of chain[i]!.fields) byName.set(f.name, f);
    }
    const result: readonly FieldMeta[] = [...byName.values()];
    cache.set(meta, result);
    return result;
}

// `ClassMeta.refs` holds OWN fields' targets only (real inheritance). An inherited
// embedded/reference/instance field's target lives in an ancestor's `refs`, so resolve a target
// name by walking the base chain leaf-first (a child entry shadows an ancestor's of the same
// name). Returns undefined when no class in the chain declares the name.
const refsCache = new WeakMap<ClassMeta, ReadonlyMap<string, ClassRef>>();

export function allRefs(meta: ClassMeta): ReadonlyMap<string, ClassRef> {
    if (meta.base === undefined) return meta.refs ?? EMPTY_REFS;
    const memo = refsCache.get(meta);
    if (memo !== undefined) return memo;

    const merged = new Map<string, ClassRef>();
    const seen = new Set<string>();
    for (let m: ClassMeta | undefined = meta; m !== undefined && !seen.has(m.name); m = m.base) {
        seen.add(m.name);
        if (m.refs === undefined) continue;
        for (const [k, v] of m.refs) if (!merged.has(k)) merged.set(k, v);
    }
    refsCache.set(meta, merged);
    return merged;
}

const EMPTY_REFS: ReadonlyMap<string, ClassRef> = new Map();

/** Resolve the target class `name` of a reference/embedded/instance type (undefined otherwise). */
export function targetOf(type: FieldType): string | undefined {
    if (type.kind === "reference" || type.kind === "embedded") return type.target;
    if (type.kind === "instance") return type.name;
    return undefined;
}
